### PR TITLE
[OTA] Add support for DefaultOTAProviders attribute

### LINKED ISSUE #13855: Support read/write of DefaultOTAProviders attribute
#### Problem
Currently, `OtaSoftwareUpdateRequestorAttrAccess` exists in ota-requestor-server to handle reads and writes of the DefaultOTAProviders attribute. However, the implementations do nothing with the list.

 #### Proposed Solution
* In `OtaSoftwareUpdateRequestorAttrAccess`, the [Write](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/ota-requestor/ota-requestor-server.cpp#L72) function should store the data somewhere.
* In `OtaSoftwareUpdateRequestorAttrAccess` the [Read](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/ota-requestor/ota-requestor-server.cpp#L57) function should read it from wherever it is stored and return the values.
* In ota-requestor-server, an API should be exposed to return the list of default OTA providers

### PR DESCRIPTION
**Pull request by carol-apple**
#### Problem
The OTA Requestor currently does not support read/write of DefaultOTAProviders attribute
Fixes: https://github.com/project-chip/connectedhomeip/issues/13855

#### Change overview
- Add server side implementation to handle the read/write (ReplaceAll operation is the only supported mode) of the DefaultOTAProviders attribute
- Cache the DefaultOTAProviders attribute values in OTARequestor class
- Select the first provider in the list as the one to use when `-q` option is passed in for a delay query on bootup
- Add AdministratorCommissioning cluster server to OTA Requestor app to support multi-fabric attribute read/write

#### Testing
- Verified basic functionality of write/read of the DefaultOTAProviders attribute works:
```
./chip-tool otasoftwareupdaterequestor write default-ota-providers '[{"fabricIndex": 1, "providerNodeID": 3735928559, "endpoint": 0}]' 0x0000001234567890 0
./chip-tool otasoftwareupdaterequestor read default-ota-providers 0x1234567890 0
```
- Verified that attributes for multiple fabrics can be written/read
```
./chip-tool pairing onnetwork-long 0x858 20202021 30 --commissioner-name beta
./chip-tool otasoftwareupdaterequestor write default-ota-providers '[{"fabricIndex": 2, "providerNodeID": 1, "endpoint": 0}]' 0x858 0 --commissioner-name beta
./chip-tool otasoftwareupdaterequestor read default-ota-providers 0x858 0 --commissioner-name beta
```
- Manual testing of basic file transfer succeeds when using the `announce-ota-provider` command
- Manual testing of basic file transfer succeeds when a valid default OTA provider attribute has been written and when used in combination with the `-q` command line option

### CODE REVIEW

File: examples/all-clusters-app/ameba/main/chipinterface.cpp [hunk 1]
```diff
@@ -97,12 +97,9 @@ AmebaOTAImageProcessor gImageProcessor;
 #endif
 
 #if CONFIG_ENABLE_OTA_REQUESTOR
-extern "C" void amebaQueryImageCmdHandler(uint32_t nodeId, uint32_t fabricId)
```
Comment (carol-apple) [inline examples/all-clusters-app/ameba/main/chipinterface.cpp:100]: @pankore I couldn't see where `amebaQueryImageCmdHandler` is being called so I removed the params to them since they aren't used. Let me know if you see an issue with this.

File: src/lib/shell/commands/Ota.cpp [hunk 2]
```diff
@@ -64,12 +52,8 @@ CHIP_ERROR NotifyImageHandler(int argc, char ** argv)
     VerifyOrReturnError(GetRequestorInstance() != nullptr, CHIP_ERROR_INCORRECT_STATE);
     VerifyOrReturnError(argc == 4, CHIP_ERROR_INVALID_ARGUMENT);
 
-    const FabricIndex fabricIndex       = static_cast<FabricIndex>(strtoul(argv[0], nullptr, 10));
```
Comment (carol-apple) [inline src/lib/shell/commands/Ota.cpp:67]: @Damian-Nordic I think this will break Notify since persistence isn't supported yet. I think between you or @hrajendran-samsung, persistence support will be added soon. Ok to remove this for now?

File: src/app/clusters/ota-requestor/OTARequestor.cpp [hunk 3]
```diff
@@ -298,11 +301,6 @@ void OTARequestor::CancelImageUpdate()
 {
     mBdxDownloader->EndDownload(CHIP_ERROR_CONNECTION_ABORTED);
 
-    // All Provider info should be invalidated
-    mProviderNodeId      = kUndefinedNodeId;
```
Comment (carol-apple) [inline src/app/clusters/ota-requestor/OTARequestor.cpp:302]: @selissia I removed this since we kind of always want `mProviderLocation` to contain something ready for the next query. We can always use the `mCurrentUpdateState` to determine if we should look at that value or not

File: src/app/clusters/ota-requestor/OTARequestor.cpp [hunk 4]
```diff
@@ -450,6 +448,69 @@ void OTARequestor::NotifyUpdateApplied(uint32_t version)
     ConnectToProvider(kNotifyUpdateApplied);
 }
 
+CHIP_ERROR OTARequestor::GetDefaultOtaProviderList(AttributeValueEncoder & encoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    return encoder.EncodeList([&](const auto & enc) -> CHIP_ERROR {
+        mDefaultOtaProviderList.ForEachActiveObject([&](ProviderLocation::Type * pl) {
```
Comment (bzbarsky-apple) [inline src/app/clusters/ota-requestor/OTARequestor.cpp:455]:
```suggestion
    return encoder.EncodeList([&](const auto & enc) -> CHIP_ERROR {
            CHIP_ERROR err = CHIP_NO_ERROR;
            mDefaultOtaProviderList.ForEachActiveObject([&](ProviderLocation::Type * pl) {
```

File: src/app/clusters/ota-requestor/OTARequestor.cpp [hunk 5]
```diff
@@ -450,6 +448,69 @@ void OTARequestor::NotifyUpdateApplied(uint32_t version)
...
+            mDefaultOtaProviderList.ReleaseObject(pl);
+        }
+        return Loop::Continue;
+    });
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR OTARequestor::AddDefaultOtaProvider(FabricIndex fabricIndex, ProviderLocation::Type const & providerLocation)
+{
+    if (providerLocation.GetFabricIndex() == kUndefinedFabricIndex)
```
Comment (bzbarsky-apple) [inline src/app/clusters/ota-requestor/OTARequestor.cpp:479]: We should be enforcing this in the IM code.  If we're not, please file an issue on that?

File: src/app/clusters/ota-requestor/OTARequestor.cpp [hunk 6]
```diff
@@ -450,6 +448,69 @@ void OTARequestor::NotifyUpdateApplied(uint32_t version)
...
+        return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+    }
+
+    // Look for an entry with the same fabric index indicated
+    bool fabricIndexExists = false;
+    mDefaultOtaProviderList.ForEachActiveObject([&](auto * pl) {
+        if (pl->GetFabricIndex() == fabricIndex)
+        {
+            fabricIndexExists = true;
+            return Loop::Break;
```
Comment (bzbarsky-apple) [inline src/app/clusters/ota-requestor/OTARequestor.cpp:490]: Fwiw, could just use the return value of `ForEachActiveObject` to see whether the loop was terminated by Loop::Break or completed, instead of closing over a boolean.  Either way, though.

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp [hunk 7]
```diff
@@ -67,14 +71,61 @@ CHIP_ERROR OtaSoftwareUpdateRequestorAttrAccess::Write(const ConcreteDataAttribu
...
+        return requestor->ClearDefaultOtaProviderList(aDecoder.AccessingFabricIndex());
+    }
+    case ConcreteDataAttributePath::ListOperation::ReplaceItem:
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    case ConcreteDataAttributePath::ListOperation::DeleteItem:
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    case ConcreteDataAttributePath::ListOperation::AppendItem: {
+        OtaSoftwareUpdateRequestor::Structs::ProviderLocation::DecodableType item;
+        ReturnErrorOnFailure(aDecoder.Decode(item));
+        return requestor->AddDefaultOtaProvider(aDecoder.AccessingFabricIndex(), item);
```
Comment (bzbarsky-apple) [inline src/app/clusters/ota-requestor/ota-requestor-server.cpp:123]: `item` will also have a fabric index, set to `aDecoder.AccessingFabricIndex()`, right?  So could just pass `item` here....

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp [hunk 8]
```diff
@@ -67,14 +71,61 @@ CHIP_ERROR OtaSoftwareUpdateRequestorAttrAccess::Write(const ConcreteDataAttribu
...
+    chip::OTARequestorInterface * requestor = chip::GetRequestorInstance();
+    if (requestor == nullptr)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+
+    switch (aPath.mListOp)
+    {
+    case ConcreteDataAttributePath::ListOperation::ReplaceAll: {
         DataModel::DecodableList<OtaSoftwareUpdateRequestor::Structs::ProviderLocation::DecodableType> list;
```
Comment (Damian-Nordic) [inline src/app/clusters/ota-requestor/ota-requestor-server.cpp:105]: So we treat 'replace-all' as 'clear'? Is that OK?

File: src/app/clusters/ota-requestor/OTARequestor.h [hunk 9]
```diff
@@ -286,6 +291,8 @@ class OTARequestor : public OTARequestorInterface, public BDXDownloader::StateDe
     OTAUpdateStateEnum mCurrentUpdateState = OTAUpdateStateEnum::kIdle;
     Server * mServer                       = nullptr;
     chip::Optional<bool> mRequestorCanConsent;
+    BitMapObjectPool<ProviderLocationType, CHIP_CONFIG_MAX_FABRICS> mDefaultOtaProviderList;
```
Comment (kpschoedel) [inline src/app/clusters/ota-requestor/OTARequestor.h:294]: Drive-by request to use `ObjectPool` (identical to `BitMapObjectPool` on small platforms, uses heap on big platforms, which gives better ASAN coverage).

**Review by Damian-Nordic — APPROVED** [review]

File: src/app/clusters/ota-requestor/OTARequestor.cpp (hunk 3 above)
Comment (selissia) [reply src/app/clusters/ota-requestor/OTARequestor.cpp:302]: `mProviderLocation` would be empty after the initialization until the Provider info has been passed to the Requestor somehow -- so having it empty is a valid state. 
In general, I think it's always better to have empty info rather than stale info.

File: src/app/clusters/ota-requestor/OTARequestor.cpp [hunk 10]
```diff
@@ -450,6 +448,69 @@ void OTARequestor::NotifyUpdateApplied(uint32_t version)
...
+        }
+        return Loop::Continue;
+    });
+
+    // Check if there is already an entry with the same fabric index
+    if (fabricIndexExists)
```
Comment (selissia) [inline src/app/clusters/ota-requestor/OTARequestor.cpp:496]: Wouldn't it be better to just overwrite the entry? We are already saying that we only support ReplaceAll.

Comment (bzbarsky-apple) [reply src/app/clusters/ota-requestor/OTARequestor.cpp:496]: That would not be per-spec, right?

In particular, if a client tries to write a list with two entries in it, the spec behavior is to have an error, not to just silently pretend like the first entry was not there.

**Review by selissia — APPROVED** [review]

**Review by harsha-rajendran — APPROVED** [review]

File: src/app/clusters/ota-requestor/OTARequestor.cpp (hunk 10 above)
Comment (selissia) [reply src/app/clusters/ota-requestor/OTARequestor.cpp:496]: Sure, writing a list with two entries for the same fabric is invalid but is this the place to detect it? My concern is attempting to write a valid list while a fabric entry already exists in the cluster. Should the whole write operation fail just because one entry already exists?
Or will the write succeed for the rest of the entries on the list but not this one? (Which is also kinda counter-intuitive)

File: examples/all-clusters-app/ameba/main/chipinterface.cpp (hunk 1 above)
Comment (carol-apple) [reply examples/all-clusters-app/ameba/main/chipinterface.cpp:100]: For posterity, @pankore and I spoke offline and `amebaQueryImageCmdHandler` is called from the ameba SDK. The ameba SDK will adapt to this change once this PR merges.

File: src/app/clusters/ota-requestor/OTARequestor.cpp (hunk 3 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/OTARequestor.cpp:302]: Yes, the empty case is valid. But I also don't see it as stale info. When an update has ended (whether success or cancelled), we revert back to the default OTA provider. That is valid to be used for the next query.

File: src/app/clusters/ota-requestor/OTARequestor.cpp (hunk 5 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/OTARequestor.cpp:479]: I see that even if I specify an invalid fabric index on the client side, it always gets converted to the accessing fabric index. I'll remove this check.

File: src/app/clusters/ota-requestor/OTARequestor.cpp (hunk 10 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/OTARequestor.cpp:496]: Current behavior will succeed until an error is encountered (so a partial write could occur leading up to the error). I didn't want to put too much logic here but if we really wanted to, perhaps the right behavior should be to wipe out the valid entries that were written already. This would enforce that a successful write could only occur if the client passed in only valid values (in this case, only one entry). I'm open to this if you guys feel this is necessary.

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp (hunk 8 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/ota-requestor-server.cpp:105]: With chunking, this is how it's done...

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp (hunk 7 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/ota-requestor-server.cpp:123]: I didn't want to make the assumption that IM layer would make sure `item.GetFabricIndex()` is correct (set to the current access fabric index). If I could make that assumption, then yes, just passing in `item` would suffice.

File: src/app/clusters/ota-requestor/OTARequestor.cpp (hunk 10 above)
Comment (selissia) [reply src/app/clusters/ota-requestor/OTARequestor.cpp]: Is there a way to detect duplicates in the input list earlier in the processing? That way there's no need to wipe out anything. Also that way we can silently overwrite the existing entries instead of failing (which is the case I was concerned about here).

Comment (selissia) [reply src/app/clusters/ota-requestor/OTARequestor.cpp]: But I don't think any of this is hugely important, I'm OK with any choice

Comment (bzbarsky-apple) [reply src/app/clusters/ota-requestor/OTARequestor.cpp]: > Is there a way to detect duplicates in the input list earlier in the processing?

Not easily.  This code is given the list items one item at a time...

File: src/app/clusters/ota-requestor/ota-requestor-server.cpp (hunk 7 above)
Comment (bzbarsky-apple) [reply src/app/clusters/ota-requestor/ota-requestor-server.cpp:123]: You can absolutely make that assumption.

Comment (carol-apple) [reply src/app/clusters/ota-requestor/ota-requestor-server.cpp:123]: I'll pass in only `item` then

File: src/app/clusters/ota-requestor/OTARequestor.cpp (hunk 10 above)
Comment (carol-apple) [reply src/app/clusters/ota-requestor/OTARequestor.cpp]: Keeping the code as-is. Currently, we can't write more than one entry for the same accessing fabric index anyway. Any additional entries would be ignored (and an error message logged to indicate so).
